etl/src/v2/workers/table_sync.rs

```diff
@@ -0,0 +1,227 @@
+    ) -> RwLockReadGuard<'_, StateInner> {
+        loop {
+            // We grab hold of the state change notify in case we don't immediately have the state
+            // that we want.
+            let state_change = {
+                let inner = self.inner.read().await;
+
+                if inner.relation_subscription_state.status == status {
+                    return inner;
+                }
+
+                inner.state_change.clone()
+            };
+
+            // We wait for a state change.
+            state_change.notified().await;
+
+            // We read the state and return the lock to the state.
+            let inner = self.inner.read().await;
+            if inner.relation_subscription_state.status == status {
+                return inner;
+            }
+        }
+    }
+}
+
+impl Deref for TableSyncWorkerState {
+    type Target = RwLock<StateInner>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(Debug)]
+pub struct TableSyncWorkerHandle {
+    state: TableSyncWorkerState,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl WorkerHandle<TableSyncWorkerState> for TableSyncWorkerHandle {
+    fn state(&self) -> TableSyncWorkerState {
+        self.state.clone()
+    }
+
+    async fn wait(mut self) {
+        let Some(handle) = self.handle.take() else {
+            return;
+        };
+
+        // TODO: properly handle failure.
+        handle.await.expect("Table sync worker failed");
+    }
+}
+
+#[derive(Debug)]
+pub struct TableSyncWorker<S, D> {
+    state_store: S,
+    destination: D,
+    rel_id: Oid,
+}
+
+impl<S, D> TableSyncWorker<S, D> {
+    pub fn new(state_store: S, destination: D, rel_id: Oid) -> Self {
+        Self {
+            state_store,
+            destination,
+            rel_id,
+        }
+    }
+}
+
+impl<S, D> Worker<TableSyncWorkerHandle, TableSyncWorkerState> for TableSyncWorker<S, D>
+where
+    S: PipelineStateStore + Clone + Send + 'static,
+    D: Destination + Clone + Send + 'static,
+{
+    async fn start(self) -> TableSyncWorkerHandle {
+        let relation_subscription_state = self
+            .state_store
+            .load_relation_subscription_state(&self.rel_id)
+            .await;
+
+        let state = TableSyncWorkerState::new(relation_subscription_state);
+
+        let state_clone = state.clone();
+        let handle = tokio::spawn(async move {
+            // We first start syncing the table.
+            start_table_sync(
+                self.state_store.clone(),
+                self.destination.clone(),
+                state_clone,
+            )
+            .await;
+
+            // If we succeed syncing the table, we want to start the same apply loop as in the apply
+            // worker but starting from the `0/0` LSN which means that the slot is starting streaming
+            // from its consistent snapshot.
+            // TODO: check if this is the right LSN to start with.
```

Comment:
The correct LSN in this case is the consistent point returned from the `create_slot` command not 0/0.

Reply:
If you supply `0/0` Postgres should default to the consistent point since it won't have any `confirmed_flush_lsn` but this has to be checked in prod. If that doesn't happen, I will have to manually supply it.

Reply:
If `0/0` is passed, the replication will start from `restart_lsn` column in `pg_replication_slots` view, not `confirmed_flush_lsn`. `restart_lsn` value can be different (older) from the consistent point. E.g. if I compare the consistent point returned by `select pg_create_logical_replication_slot('my_slot', 'pgoutput');` against the slot's `restart_lsn` by `select restart_lsn from pg_replication_slots;` they will be different even when there are no concurrent transactions. In face of concurrent transactions when creating a slot, the difference can be even wider. This means passing `0/0` can lead to transactions appearing in CDC stream which are older than the table snapshot. In general, passing the consistent point is safer as Postgres also does that.

Reply:
I'm probably wrong that it restarts from `restart_lsn` if `0/0` is passed ([ref](https://github.com/postgres/postgres/blob/4fbb46f61271f4b7f46ecad3de608fc2f4d7d80f/src/backend/replication/walsender.c#L1482-L1490)). It starts reading from `restarts_lsn` but maybe doesn't send from that location. Try it out actually and see what happens.

Reply:
Confirmed from the docs: 

> START_REPLICATION SLOT slot_name LOGICAL XXX/XXX [ ( option_name [ option_value ] [, ...] ) ] [#](https://www.postgresql.org/docs/current/protocol-replication.html#PROTOCOL-REPLICATION-START-REPLICATION-SLOT-LOGICAL)
>
> Instructs server to start streaming WAL for logical replication, starting at either WAL location XXX/XXX or the slot's confirmed_flush_lsn (see [Section 52.19](https://www.postgresql.org/docs/current/view-pg-replication-slots.html)), whichever is greater

It should be ok then.